objects/src/assets/mod.rs

```diff
@@ -0,0 +1,453 @@
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct FungibleAsset {
+    faucet_id: AccountId,
+    amount: u64,
+}
+
+impl FungibleAsset {
+    // CONSTANTS
+    // --------------------------------------------------------------------------------------------
+    /// Specifies a maximum amount value for fungible assets which can be at most a 63-bit value.
+    pub const MAX_AMOUNT: u64 = (1_u64 << 63) - 1;
+
+    // CONSTRUCTOR
+    // --------------------------------------------------------------------------------------------
+    /// Returns a fungible asset instantiated with the provided faucet ID and amount.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - The provided faucet ID is not for a fungible asset faucet.
+    /// - The provide amount is greater than 2^63 - 1.
+    pub fn new(faucet_id: AccountId, amount: u64) -> Result<Self, AssetError> {
+        if !faucet_id.is_fungible_faucet() {
+            return Err(AssetError::not_a_fungible_faucet_id(faucet_id));
+        }
+
+        // construct the asset and make sure it passes the validation logic
+        let asset = Self { faucet_id, amount };
+        asset.validate()?;
+        Ok(asset)
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Return ID of the faucet which issued this asset.
+    pub fn faucet_id(&self) -> AccountId {
+        self.faucet_id
+    }
+
+    /// Returns the amount of this asset.
+    pub fn amount(&self) -> u64 {
+        self.amount
+    }
+
+    // OPERATIONS
+    // --------------------------------------------------------------------------------------------
+
+    /// Adds two fungible assets together and returns the result.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - The assets were not issued by the same faucet.
+    /// - The total value of assets is greater than or equal to 2^63.
+    #[allow(clippy::should_implement_trait)]
+    pub fn add(self, other: Self) -> Result<Self, AssetError> {
+        if self.faucet_id != other.faucet_id {
+            return Err(AssetError::inconsistent_faucet_ids(self.faucet_id, other.faucet_id));
+        }
+
+        let amount = self.amount.checked_add(other.amount).expect("overflow!");
+        if amount > Self::MAX_AMOUNT {
+            return Err(AssetError::amount_too_big(amount));
+        }
+
+        Ok(Self {
+            faucet_id: self.faucet_id,
+            amount,
+        })
+    }
+
+    /// Subtracts the specified amount from this asset and returns the resulting asset.
+    ///
+    /// # Errors
+    /// Returns an error if this asset's amount is smaller than the requested amount.
+    pub fn sub(&mut self, amount: u64) -> Result<Self, AssetError> {
+        self.amount = self
+            .amount
+            .checked_sub(amount)
+            .ok_or(AssetError::AssetAmountNotSufficient(self.amount, amount))?;
+
+        self.amount -= amount;
```

Comment:
We should delete this line now since the subtraction is being performed above

Reply:
Ah yes! Totally missed this. Fixed!